llvm/lib/CodeGen/MachineOutliner.cpp

```diff
@@ -1075,6 +1076,23 @@ bool MachineOutliner::outline(
                       << " B) > threshold (" << OutlinerBenefitThreshold
                       << " B)\n");
 
+    // Remove all Linker Optimization Hints from the candidates since we did not
+    // check if the set of hints are the same for each of them.
+    // TODO: The intersection of the LOHs from all candidates should be legal in
+    // the outlined function.
+    SmallPtrSet<MachineInstr *, 2> MIs;
+    std::optional<size_t> MinRemovedLOHs;
+    for (Candidate &C : OF->Candidates) {
+      const TargetInstrInfo &TII = *C.getMF()->getSubtarget().getInstrInfo();
```

Comment:
Should this really be a subtarget property in an outlining context? This would need to be module level / TargetMachine?

Reply:
My point was more this should not go into TargetInstrInfo. What happens if the subtargets are different?

Reply:
I think it's actually safe because arbitrarily choose the subtarget from the first function when creating the outlined function.

https://github.com/llvm/llvm-project/blob/2ee8fdbfddcca86ac079104718e6fda3aabed0eb/llvm/lib/CodeGen/MachineOutliner.cpp#L829-L832

I think we could use `TargetMachine`, but that would require adding a dependency on `TargetPassConfig`. I would rather try to first move the MachineOutliner to use `TargetMachine`, then try to fix this case.

https://github.com/llvm/llvm-project/blob/2ee8fdbfddcca86ac079104718e6fda3aabed0eb/llvm/lib/CodeGen/MachinePipeliner.cpp#L612-L621

Reply:
Just moved it to TargetMachine anyway